### PR TITLE
Add support of non existent in a timeframe data-points

### PR DESCRIPTION
**Pull request by vovakulikov**
Closes https://github.com/sourcegraph/sourcegraph/issues/20772

**Important** This PR should be merged together with https://github.com/sourcegraph/sourcegraph-search-insights/pull/5

If user set up their insights in a way to get a timeframe when repositories in the insight repositories field hadn't existed It breaks the insight and shows nothing at all but the error message. (see screenshot below)

Now we just skip points which we can't fetch because repositories haven't existed at that moment

| BEFORE             |  AFTER |
:-------------------------:|:-------------------------:
 image | image

### CODE REVIEW

File: client/web/src/insights/core/backend/api/get-search-insight-content.ts [hunk 1]
```diff
@@ -4,27 +4,56 @@ import { defer } from 'rxjs'
...
 export async function getSearchInsightContent(insight: SearchInsightSettings): Promise<LineChartContent<any, string>> {
     const step = insight.step || { days: 1 }
     const { repositories: repos } = insight
     const dates = getDaysToQuery(step)
 
-    // Get commits to search for each day
+    // -------- Initialize data ---------
+    const data: InsightSeriesData[] = []
+
+    for (const date of dates) {
```
Comment (vovakulikov) [inline client/web/src/insights/core/backend/api/get-search-insight-content.ts:36]: Fill in result data array by empty points

**Comment by felixfbecker** [comment]
I see there is a pattern rendered when the repo didn't exist (which looks super cool!). Two questions in my mind:
1. What happens if there are multiple series, and only one of the series didn't exist prior to a certain point in time, but the other series did exist at that point? I assume we wouldn't render the pattern them, but the series would simply not have a data point yet?
2. What if one series had multiple repos, and only one of the repos wasn't available yet? I assume it would show the count from only the repo that existed?
I'll let someone else review the code :)

**Comment by vovakulikov** [comment]
@felixfbecker Nice call! 

1. You actually can see that case on screenshot above. The purple series has data the blue one doesn't. In this case, we just set series which doesn't have any values for that period to zero. Just because we still want to show that series to be able to compare all series when one of the series still has data. 

image

2. That's correct) 
> I assume it would show the count from only the repo that existed?

**Comment by felixfbecker** [comment]
I'm confused, in that screenshot it seems that blue and purple have data points for exactly the same  X values. At which X value are you saying does the blue one not have a data point where purple has one?

**Review by eseliger — APPROVED** [review]
Code LGTM

**Comment by vovakulikov** [comment]
@felixfbecker 
>At which X value are you saying does the blue one not have a data point where purple has one?

The first two points of the blue line don't have values (these two moments in an insight timeline when we didn't have any results for this series) 

As a result, if there are multiple series, and only one of the series didn't exist prior to a certain point in time, but the other series did exist at that point we set zero values to the points of series that didn't exist

**Comment by vovakulikov** [comment]
Since we don't have any new comments and issues I'm about to merge this. @felixfbecker @eseliger thanks for reviewing it.

**Comment by felixfbecker** [comment]
I'm curious how the chart decides to render the pattern at the beginning. My understanding was that the chart's X range is determined by the minimum and maximum X value. So how would there be a situation in which that pattern is rendered? Or in other words, how does it decide the range of the X axis, if not by the min and max X values?

**Comment by vovakulikov** [comment]
@felixfbecker here is a small part of logic which responsible for that decision.

```ts
 for (const line of series) {            
   const lineKey = line.dataKey
   const lineYAccessor = accessors.y[lineKey]            
   
   firstNonNullablePoints.push(getFirstNonNullablePoint(data, lineYAccessor))
}
 ```
 
If there is a datum with a null value for one of the series this is our sign that we have to render non active background on the chart. Important note that we pick the first such datum. 
 
https://github.com/sourcegraph/sourcegraph/pull/21288/files#diff-ceab1cf0b38fb27e97c8ec502c4d739a1e3169157941b4b28cc6c3b17014b3bbR48-R53

Also, I added a story for this case just to show an example of how it looks like and what is the data for line chart behind this case 

https://github.com/sourcegraph/sourcegraph/pull/21288/files#diff-93f24791de8be69b01503ccd7d22d8adcac080e729ce2a00ad59a1d893b14de8R78

**Comment by felixfbecker** [comment]
Ah, so the extension returns `null` as the value for the areas that should have the pattern? It would be good to document that in the API in sourcegraph.d.ts, perhaps even make it a union type `| null`.

Does it break old versions of Sourcegraph if the new version of the extension returns `null` as the value?

**Comment by vovakulikov** [comment]
@felixfbecker 
>Ah, so the extension returns null as the value for the areas that should have the pattern? It would be good to document that in the API in sourcegraph.d.ts, perhaps even make it a union type | null.

Extension set null value for each series which doesn't have value for a particular point of time. As I see we can't add type for this because data for each series placed within datum object which is generic in our codebase. (see data field below)

```
 export interface LineChartContent<D extends object, XK extends keyof D> {
        chart: 'line'

        /** An array of data objects, with one element for each step on the X-axis. */
        data: D[]

        /** The series (lines) of the chart. */
        series: {}[]

        xAxis: ChartAxis<XK, D>
    }
```
>Does it break old versions of Sourcegraph (i.e. before this PR) if the new version of the extension returns null as the value?

This is actually a good call. I will check later but at first glance, it shouldn't break anything, the null value is a valid data point for our charts but I'm not sure about old old charts (re-chart library) I believe we have these charts in 3.26

**Comment by vovakulikov** [comment]
@felixfbecker I suggest we change the place for this extension API conversation to this PR's thread https://github.com/sourcegraph/sourcegraph-search-insights/pull/5

**Pull request merged**